src/Dapr.PluggableComponents.AspNetCore/DaprPluggableComponentsApplication.cs

```diff
@@ -0,0 +1,169 @@
+using System.Collections.Concurrent;
+using Dapr.PluggableComponents.Adaptors;
+using Dapr.PluggableComponents.Components.StateStore;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Dapr.PluggableComponents;
+
+public sealed class DaprPluggableComponentsApplication
+{
+    public static DaprPluggableComponentsApplication Create()
+    {
+        return Create(new DaprPluggableComponentsApplicationOptions());
+    }
+
+    public static DaprPluggableComponentsApplication Create(string socketName)
+    {
+        return Create(new DaprPluggableComponentsApplicationOptions { SocketName = socketName });
+    }
+
+    public static DaprPluggableComponentsApplication Create(DaprPluggableComponentsApplicationOptions options)
+    {
+        return new DaprPluggableComponentsApplication(options);
+    }
+
+    private readonly DaprPluggableComponentsApplicationOptions options;
+
+    private readonly ConcurrentBag<Action<WebApplicationBuilder>> builderActions = new ConcurrentBag<Action<WebApplicationBuilder>>();
+    private readonly ConcurrentBag<Action<WebApplication>> appActions = new ConcurrentBag<Action<WebApplication>>();
+
+    private DaprPluggableComponentsApplication(DaprPluggableComponentsApplicationOptions options)
+    {       
+        this.options = options;
+    }
+
+    #region State Store Members
+
+    public void AddStateStore<TStateStore>(Func<IServiceProvider, string?, TStateStore> stateStoreFactory)
```

Comment:
What is the `string` parameter of this factory func? Is this the related `instance` ? If yes, it is not clear to me if its being automatically multiplexed by the sdk or if it is a way to let the user do whatever they want with the instance id (including handle multiplexing by their own)

Reply:
The string here is the Dapr instance ID, pulled from the gRPC headers, to allow developers to create an instance of a component per Dapr instance.  "Our" infrastructure is handling the multiplexing in that this factory method is called only once per Dapr instance, but provides the ability for developers to know when and for what an instance is being created, whether for logging or other purposes.

Alternatively, if we thought the instance ID wasn't useful, we *may* be able to manage the instance creation ourself (though it's not a singleton or request-scoped value, so I'm not sure how easily we can create arbitrary instances that still allow proper service provider-based DI) and then make singleton vs. instance-based just a boolean/enumerated option during registration.

Reply:
I'd prefer to hide this complexity and just multiplex by default.

Reply:
ASP.NET DI has singleton, scoped (per-server-request), and transient (per-component-request) DI scopes and I don't think any of those quite match what we need.  Potentially the last scope *might* work, but the problem is that we don't own the type being registered and so we can't control what other types might inject it, which could cause issues in the developer's application if they're not fully cognizant of its use and its implications.  I don't see a lot of that latter scope in practise (likely because its more difficult to reason about in a large mesh of components).

Note that the latest revision moves the arguments into a single context object, which hides some of that complexity; the developer can completely ignore it if their type doesn't need it.

Simplest use (state store doesn't care about instance ID or need other registered services):

```csharp
app.RegisterStateStore(_ => new MyStateStore());
```

Use case when developer needs access to registered services:

```csharp
app.RegisterStateStore(context => new MyStateStore(context.ServiceProvider.GetRequiredService<MyDependency>()));
```